In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb

# <u> 1. GATHERING OF DATA

In [382]:
transac = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='Transactions',header = 1)
transac.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [15]:
cusdem = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='CustomerDemographic', header = 1)
cusdem.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [16]:
cusadd = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='CustomerAddress', header = 1)
cusadd.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

######  NOTE: Let's hold out the new customer list for now.

# <U> 2. ASSESMENT OF DATA 

##### Transaction data

In [122]:
transac.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

In [8]:
transac.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
transaction_id             20000 non-null int64
product_id                 20000 non-null int64
customer_id                20000 non-null int64
transaction_date           20000 non-null datetime64[ns]
online_order               19640 non-null float64
order_status               20000 non-null object
brand                      19803 non-null object
product_line               19803 non-null object
product_class              19803 non-null object
product_size               19803 non-null object
list_price                 20000 non-null float64
standard_cost              19803 non-null float64
product_first_sold_date    19803 non-null float64
dtypes: datetime64[ns](1), float64(4), int64(3), object(5)
memory usage: 2.0+ MB


In [383]:
ad = transac.isnull().sum() #checking for missing values in our dataset
ad

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

In [170]:
ad/ len(transac)

transaction_id             0.00000
product_id                 0.00000
customer_id                0.00000
transaction_date           0.00000
online_order               0.01800
order_status               0.00000
brand                      0.00985
product_line               0.00985
product_class              0.00985
product_size               0.00985
list_price                 0.00000
standard_cost              0.00985
product_first_sold_date    0.00985
dtype: float64

In [157]:
temp = transac[transac.isnull().any(axis=1)]
temp

transaction_id  product_id  customer_id transaction_date  online_order  \
97                 98          49          333       2017-06-23           NaN   
136               137           0          431       2017-09-23           0.0   
159               160           0         3300       2017-08-27           0.0   
166               167          90         3177       2017-04-26           NaN   
169               170           6          404       2017-10-16           NaN   
250               251          63         1967       2017-04-11           NaN   
300               301          78         2530       2017-03-24           NaN   
336               337          82         1615       2017-10-30           NaN   
342               343          61         1478       2017-03-11           NaN   
366               367           0         1614       2017-03-10           0.0   
398               399          83         1306       2017-12-18           NaN   
406               407           0         2559       2017-06-14           1.0   
476               477          73          367       2017-02-12           NaN   
529               530          62         1202       2017-03-31           NaN   
558               559          20         3104       2017-03-21           NaN   
576               577           1          149       2017-05-04           NaN   
676               677           0         2609       2017-07-02           0.0   
760               761           4         2428       2017-12-20           NaN   
769               770           4         2414       2017-09-10           NaN   
780               781           0          897       2017-05-10           1.0   
843               844          40         3412       2017-07-09           NaN   
914               915          39         2756       2017-03-16           NaN   
917               918          35         2281       2017-04-30           NaN   
1003             1004           0         2254       2017-05-16           0.0   
1060             1061          63         1342       2017-07-14           NaN   
1071             1072          36         1505       2017-11-09           NaN   
1130             1131           0         2163       2017-03-09           0.0   
1159             1160          73         1722       2017-12-15           NaN   
1192             1193          71           79       2017-11-17           NaN   
1196             1197           0         2234       2017-04-04           0.0   
...               ...         ...          ...              ...           ...   
18819           18820           0         1503       2017-03-03           0.0   
18821           18822          30         3464       2017-12-13           NaN   
18846           18847          57           50       2017-09-20           NaN   
18982           18983          21          743       2017-01-03           NaN   
19017           19018           0         3384       2017-12-01           0.0   
19024           19025          36         1937       2017-08-04           NaN   
19045           19046           0         2410       2017-11-02           1.0   
19082           19083          90         1204       2017-12-06           NaN   
19100           19101           8         2713       2017-10-21           NaN   
19132           19133           0          181       2017-05-04           1.0   
19176           19177           0         1043       2017-06-06           0.0   
19196           19197           0         1182       2017-12-06           1.0   
19205           19206           0         2122       2017-01-04           1.0   
19271           19272          91         3017       2017-07-07           NaN   
19282           19283          55         1364       2017-08-02           NaN   
19303           19304          35         1776       2017-08-22           NaN   
19340           19341           0          443       2017-12-26           1.0   
19368           19369          66     

In [159]:
temp[temp['product_id']==0]


transaction_id  product_id  customer_id transaction_date  online_order  \
136               137           0          431       2017-09-23           0.0   
159               160           0         3300       2017-08-27           0.0   
366               367           0         1614       2017-03-10           0.0   
406               407           0         2559       2017-06-14           1.0   
676               677           0         2609       2017-07-02           0.0   
780               781           0          897       2017-05-10           1.0   
1003             1004           0         2254       2017-05-16           0.0   
1130             1131           0         2163       2017-03-09           0.0   
1196             1197           0         2234       2017-04-04           0.0   
1282             1283           0          397       2017-08-09           0.0   
1621             1622           0          736       2017-09-07           NaN   
1653             1654           0         1972       2017-10-05           0.0   
1680             1681           0         2733       2017-05-18           0.0   
1752             1753           0          925       2017-07-25           0.0   
1778             1779           0         2153       2017-03-23           0.0   
2186             2187           0         1568       2017-03-14           1.0   
2302             2303           0         2465       2017-04-22           0.0   
2772             2773           0         1136       2017-05-08           NaN   
2947             2948           0         3212       2017-09-14           0.0   
2990             2991           0          812       2017-04-10           1.0   
3194             3195           0         1902       2017-12-09           0.0   
3259             3260           0         3168       2017-02-24           1.0   
3408             3409           0           52       2017-03-11           1.0   
3781             3782           0         1503       2017-11-26           1.0   
3877             3878           0          696       2017-07-30           0.0   
3904             3905           0         3323       2017-03-11           NaN   
3929             3930           0         3415       2017-10-30           NaN   
4024             4025           0         3264       2017-05-05           1.0   
4073             4074           0         1841       2017-04-09           0.0   
4127             4128           0         3242       2017-01-04           1.0   
...               ...         ...          ...              ...           ...   
17235           17236           0         1026       2017-11-16           1.0   
17291           17292           0          171       2017-05-30           1.0   
17331           17332           0         1702       2017-02-20           1.0   
17431           17432           0         2135       2017-08-15           1.0   
17534           17535           0         1210       2017-07-29           1.0   
17767           17768           0          168       2017-03-13           0.0   
17918           17919           0         1295       2017-11-10           1.0   
17987           17988           0         2720       2017-03-11           0.0   
18021           18022           0         3104       2017-05-21           0.0   
18051           18052           0          918       2017-06-22           1.0   
18244           18245           0         3261       2017-05-17           1.0   
18277           18278           0          586       2017-02-20           1.0   
18279           18280           0          916       2017-09-17           0.0   
18344           18345           0         3359       2017-06-06           0.0   
18492           18493           0          331       2017-02-17           1.0   
18543           18544           0         1078       2017-01-15           1.0   
18553           18554           0         1416       2017-10-24           1.0   
18778           18779           0     

In [160]:
print(temp['product_id'].value_counts())

0     221
83      8
63      8
64      8
21      7
35      7
36      7
90      7
1       6
4       6
20      6
41      6
43      6
78      6
49      6
95      6
89      6
84      5
46      5
40      5
68      5
32      5
73      5
55      5
6       5
3       5
22      4
15      4
8       4
56      4
     ... 
79      2
59      2
54      2
14      2
16      2
18      2
37      2
67      2
87      2
58      2
93      2
65      2
31      2
30      2
19      2
44      2
26      2
25      2
77      1
29      1
13      1
12      1
45      1
53      1
5       1
70      1
81      1
75      1
97      1
60      1
Name: product_id, Length: 96, dtype: int64


In [166]:
transac[transac['product_id']==2].sample(5)


transaction_id  product_id  customer_id transaction_date  online_order  \
18894           18895           2          197       2017-05-14           1.0   
6532             6533           2         2532       2017-08-12           0.0   
19887           19888           2         1686       2017-05-10           1.0   
6293             6294           2          627       2017-11-02           0.0   
3927             3928           2         2410       2017-02-04           1.0   

      order_status           brand product_line product_class product_size  \
18894     Approved  Giant Bicycles         Road           low        small   
6532      Approved           Solex     Standard        medium       medium   
19887     Approved           Solex     Standard        medium       medium   
6293      Approved  Giant Bicycles         Road           low        small   
3927      Approved           Solex     Standard        medium       medium   

       list_price  standard_cost  product_first_sold_date  
18894      590.26         525.33                  40487.0  
6532        71.49          53.62                  41245.0  
19887       71.49          53.62                  38573.0  
6293       590.26         525.33                  40487.0  
3927        71.49          53.62                  41245.0

In [164]:
transac[transac['product_id']==5].sample(5)

transaction_id  product_id  customer_id transaction_date  online_order  \
4564             4565           5           10       2017-08-26           0.0   
7054             7055           5         2915       2017-02-24           0.0   
19818           19819           5          116       2017-08-16           0.0   
11594           11595           5         2485       2017-12-26           0.0   
12655           12656           5          261       2017-02-27           1.0   

      order_status          brand product_line product_class product_size  \
4564      Approved  Trek Bicycles     Mountain           low       medium   
7054      Approved  Trek Bicycles     Mountain           low       medium   
19818     Approved  Trek Bicycles     Mountain           low       medium   
11594     Approved  Trek Bicycles     Mountain           low       medium   
12655     Approved  Trek Bicycles     Mountain           low       medium   

       list_price  standard_cost  product_first_sold_date  
4564       574.64         459.71                  38258.0  
7054       574.64         459.71                  40784.0  
19818      574.64         459.71                  38258.0  
11594      574.64         459.71                  37659.0  
12655      574.64         459.71                  40784.0

In [29]:
transac.duplicated().any() 

False

In [33]:
transac.describe()

transaction_id   product_id   customer_id  online_order    list_price  \
count    20000.000000  20000.00000  20000.000000  19640.000000  20000.000000   
mean     10000.500000     45.36465   1738.246050      0.500458   1107.829449   
std       5773.647028     30.75359   1011.951046      0.500013    582.825242   
min          1.000000      0.00000      1.000000      0.000000     12.010000   
25%       5000.750000     18.00000    857.750000      0.000000    575.270000   
50%      10000.500000     44.00000   1736.000000      1.000000   1163.890000   
75%      15000.250000     72.00000   2613.000000      1.000000   1635.300000   
max      20000.000000    100.00000   5034.000000      1.000000   2091.470000   

       standard_cost  product_first_sold_date  
count   19803.000000             19803.000000  
mean      556.046951             38199.776549  
std       405.955660              2875.201110  
min         7.210000             33259.000000  
25%       215.140000             35667.000000  
50%       507.580000             38216.000000  
75%       795.100000             40672.000000  
max      1759.850000             42710.000000

In [154]:
transac.sample(10,random_state = 100)

transaction_id  product_id  customer_id transaction_date  online_order  \
2917             2918          35          299       2017-07-02           0.0   
2234             2235          62         2064       2017-03-09           0.0   
14396           14397          92         2181       2017-05-19           0.0   
1781             1782          40          979       2017-10-12           1.0   
11102           11103          17           48       2017-08-03           1.0   
732               733          50          679       2017-05-12           1.0   
7692             7693           0          229       2017-10-31           1.0   
10589           10590           6         1699       2017-06-01           1.0   
16098           16099          66         1245       2017-10-20           1.0   
4920             4921          96          490       2017-03-08           0.0   

      order_status           brand product_line product_class product_size  \
2917      Approved   Trek Bicycles     Standard           low       medium   
2234      Approved           Solex     Standard        medium       medium   
14396     Approved        WeareA2B     Standard        medium        small   
1781      Approved   Trek Bicycles         Road        medium        large   
11102     Approved           Solex     Standard          high       medium   
732       Approved        WeareA2B     Standard        medium        small   
7692      Approved  Norco Bicycles     Standard           low       medium   
10589     Approved      OHM Cycles     Standard          high       medium   
16098     Approved  Giant Bicycles         Road           low        small   
4920      Approved        WeareA2B         Road           low        small   

       list_price  standard_cost  product_first_sold_date  
2917      1057.51         154.40                  40336.0  
2234       478.16         298.72                  40487.0  
14396     1415.01        1259.36                  34556.0  
1781      1894.19         598.76                  37823.0  
11102     1024.66         614.80                  34244.0  
732        175.89         131.92                  37668.0  
7692       363.01         290.41                  38482.0  
10589      227.88         136.73                  39298.0  
16098      590.26         525.33                  40487.0  
4920      1172.78        1043.77                  37539.0

#####  Customer Demographic

In [125]:
cusdem.sample(10, random_state =10)

customer_id first_name   last_name  gender  \
2177         2178     Veriee    Colthard  Female   
2382         2383      Zaria       Eberz  Female   
1715         1716        Lou      Yellep  Female   
2303         2304     Shelia      Bellew  Female   
822           823     Stesha   Morecombe  Female   
1055         1056  Kendricks      Markel    Male   
2429         2430    Humbert      Dikles    Male   
237           238        Max  Bilborough    Male   
908           909    Suzette    Surgison  Female   
1243         1244      Glenn      Tinham       U   

      past_3_years_bike_related_purchases        DOB  \
2177                                   20 1954-03-18   
2382                                   34 1978-09-15   
1715                                   42 1987-01-30   
2303                                   65 1957-06-01   
822                                    91 1990-02-17   
1055                                   96 1966-12-18   
2429                                   38 1968-09-06   
237                                    54 1984-02-04   
908                                    77 1960-01-04   
1243                                   80        NaT   

                        job_title job_industry_category     wealth_segment  \
2177  Mechanical Systems Engineer                Retail      Mass Customer   
2382         Compensation Analyst    Financial Services      Mass Customer   
1715             Senior Developer    Telecommunications  Affluent Customer   
2303                          NaN                   NaN      Mass Customer   
822              Internal Auditor         Manufacturing  Affluent Customer   
1055    Software Test Engineer IV         Manufacturing  Affluent Customer   
2429             Quality Engineer                   NaN      Mass Customer   
237                Programmer III                   NaN      Mass Customer   
908                      Operator                   NaN     High Net Worth   
1243            Financial Analyst    Financial Services      Mass Customer   

     deceased_indicator                                            default  \
2177                  N  0ï¸â£ 1ï¸â£ 2ï¸â£ 3ï¸â£ 4ï¸â£ 5ï¸â£ 6ï¸â£ 7ï¸â...   
2382                  N                       <script>alert('hi')</script>   
1715                  N                                                nil   
2303                  N                                2018-02-01 00:00:00   
822                   N                                     ¸ËÃâıËï˿   
1055                  N                                                1/0   
2429                  N             ×Ö¸×ְתָ×testاÙصÙحات اÙتÙØ­ÙÙ   
237                   N                                           â©testâ©   
908                   N                                                 ⢠  
1243                  N                                                NaN   

     owns_car  tenure  
2177       No    12.0  
2382       No    15.0  
1715       No     5.0  
2303      Yes    11.0  
822        No     9.0  
1055       No     6.0  
2429       No     5.0  
237       Yes    14.0  
908       Yes     8.0  
1243      Yes     NaN

In [19]:
cusdem.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
customer_id                            4000 non-null int64
first_name                             4000 non-null object
last_name                              3875 non-null object
gender                                 4000 non-null object
past_3_years_bike_related_purchases    4000 non-null int64
DOB                                    3913 non-null datetime64[ns]
job_title                              3494 non-null object
job_industry_category                  3344 non-null object
wealth_segment                         4000 non-null object
deceased_indicator                     4000 non-null object
default                                3698 non-null object
owns_car                               4000 non-null object
tenure                                 3913 non-null float64
dtypes: datetime64[ns](1), float64(1), int64(2), object(9)
memory usage: 406.3+ KB


In [120]:
cusdem.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

In [126]:
temp = cusdem[cusdem.isnull().any(axis=1)]
temp

customer_id      first_name    last_name  gender  \
3               4          Talbot          NaN    Male   
4               5  Sheila-kathryn       Calton  Female   
5               6            Curr    Duckhouse    Male   
6               7            Fina       Merali  Female   
7               8             Rod        Inder    Male   
10             11           Uriah       Bisatt    Male   
15             16          Harlin         Parr    Male   
16             17           Heath      Faraday    Male   
17             18          Marjie      Neasham  Female   
21             22         Deeanne     Durtnell  Female   
22             23            Olav        Polak    Male   
29             30         Darrick    Helleckas    Male   
32             33           Ernst        Hacon    Male   
33             34        Jephthah     Bachmann       U   
35             36         Lurette     Stonnell  Female   
45             46           Kaila        Allin  Female   
47             48        Rebbecca       Casone  Female   
48             49           Nolly    Ownsworth    Male   
51             52          Curran      Bentson    Male   
56             57            Abba      Masedon       M   
58             59            Niki    Heathcote    Male   
59             60          Nadiya   Champerlen  Female   
61             62          Sorcha      Roggers  Female   
66             67          Vernon          NaN    Male   
67             68          Dahlia       Eddoes  Female   
68             69           Heidi       Milner  Female   
72             73         Minette      Worters  Female   
73             74           Pansy       Kiddie  Female   
80             81             Bee    Blazewicz  Female   
83             84            Rich    Mathiasen    Male   
...           ...             ...          ...     ...   
3929         3930         Blancha        Baldi  Female   
3930         3931           Kylie        Epine       U   
3932         3933         Chiarra         Cops  Female   
3934         3935          Teodor    Alfonsini       U   
3938         3939         Georges    Dumbelton    Male   
3944         3945         Lazarus      Donaghy    Male   
3945         3946           Wylie  FitzGilbert    Male   
3946         3947       Tanitansy    McTrustam  Female   
3950         3951          Ephrem     Hollerin    Male   
3951         3952              Di     Borsnall  Female   
3956         3957         Bernice  Scotchforth  Female   
3958         3959          Dannie       Sowray    Male   
3959         3960          Hobart       Burgan    Male   
3961         3962          Benoit          NaN    Male   
3962         3963         Ardelle       Dasent  Female   
3963         3964          Aubrey       Souten  Female   
3964         3965        Susannah       Dumphy  Female   
3965         3966          Astrix      Sigward  Female   
3967         3968       Alexandra        Kroch  Female   
3970         3971         Analise       Asbrey  Female   
3971         3972       Maribelle     Schaffel  Female   
3973         3974           Misha      Ranklin  Female   
3975         3976          Gretel     Chrystal  Female   
3978         3979           Kleon         Adam    Male   
3982         3983          Jarred        Lyste    Male   
3985         3986          Oralle     Petrello  Female   
3986         3987          Beckie      Wakeham  Female   
3997         3998          Sarene      Woolley       U   
3998         3999       Patrizius          NaN    Male   
3999         4000           Kippy      Oldland    Male   

      past_3_years_bike_related_purchases        DOB  \
3                                      33 1961-10-03   
4                                      56 1977-05-13   
5                                      35 1966-09-16   
6                                       6 1976-02-23   
7                                      31 1962-03-30   
10                                     99 1954-04-30   
15               

In [161]:
cusdem.gender.value_counts()

Female    2037
Male      1872
U           88
M            1
Femal        1
F            1
Name: gender, dtype: int64

In [130]:
cusdem.DOB.sort_values(ascending = True)

33     1843-12-21
719    1931-10-23
1091   1935-08-22
3409   1940-09-22
2412   1943-08-11
657    1944-01-24
1242   1953-08-09
1564   1953-08-17
1176   1953-08-31
3470   1953-09-05
2812   1953-09-10
2018   1953-09-13
1904   1953-09-17
559    1953-10-11
1914   1953-10-12
0      1953-10-12
3102   1953-10-21
2745   1953-10-25
3258   1953-11-04
2494   1953-11-12
424    1953-11-19
2958   1953-11-23
3624   1953-11-30
2481   1953-12-10
718    1953-12-30
466    1954-01-02
365    1954-01-08
254    1954-01-15
2      1954-01-20
3369   1954-01-22
          ...    
2695          NaT
2696          NaT
2853          NaT
2919          NaT
2962          NaT
2998          NaT
3011          NaT
3085          NaT
3150          NaT
3221          NaT
3222          NaT
3254          NaT
3287          NaT
3297          NaT
3311          NaT
3321          NaT
3342          NaT
3364          NaT
3472          NaT
3509          NaT
3512          NaT
3564          NaT
3653          NaT
3717          NaT
3726      

In [131]:
cusdem.describe()

customer_id  past_3_years_bike_related_purchases       tenure
count  4000.000000                          4000.000000  3913.000000
mean   2000.500000                            48.890000    10.657041
std    1154.844867                            28.715005     5.660146
min       1.000000                             0.000000     1.000000
25%    1000.750000                            24.000000     6.000000
50%    2000.500000                            48.000000    11.000000
75%    3000.250000                            73.000000    15.000000
max    4000.000000                            99.000000    22.000000

In [137]:
cusdem.duplicated().any() 

False

##### Customer Address

In [141]:
cusadd.sample(20,random_state=30)

customer_id                   address  postcode            state  \
1743         1748         09810 Dayton Hill      4035              QLD   
1241         1246      65209 Merrick Avenue      4873              QLD   
3233         3238        8 Rockefeller Lane      2306              NSW   
856           861          55 Hovde Parkway      3975              VIC   
3823         3828         0964 Bayside Park      3136              VIC   
1729         1734         0 Westport Center      4558              QLD   
3215         3220         580 Blaine Center      2029              NSW   
3907         3912         55 Anzinger Drive      4152              QLD   
3070         3075           08180 Golf Park      3068              VIC   
3968         3973           4 Express Plaza      2191              NSW   
3465         3470      176 Sutteridge Alley      3121              VIC   
40             45       2028 Lakewood Place      2480  New South Wales   
3362         3367     3 Weeping Birch Plaza      3677              VIC   
3375         3380            2 Roxbury Pass      3977              VIC   
3982         3987            9 Grover Point      4218              QLD   
417           422             509 Fisk Hill      2031              NSW   
1288         1293      378 Butterfield Pass      4300              QLD   
2763         2768        9 Cascade Junction      2016              NSW   
2587         2592       303 Michigan Avenue      3977              VIC   
3415         3420  00283 Loftsgordon Center      4118              QLD   

        country  property_valuation  
1743  Australia                   8  
1241  Australia                   2  
3233  Australia                   4  
856   Australia                   8  
3823  Australia                   9  
1729  Australia                   4  
3215  Australia                  10  
3907  Australia                   9  
3070  Australia                   8  
3968  Australia                  10  
3465  Australia                  11  
40    Australia                   3  
3362  Australia                   2  
3375  Australia                   5  
3982  Australia                  10  
417   Australia                  11  
1288  Australia                   6  
2763  Australia                  12  
2587  Australia                   5  
3415  Australia                   5

In [142]:
cusadd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
customer_id           3999 non-null int64
address               3999 non-null object
postcode              3999 non-null int64
state                 3999 non-null object
country               3999 non-null object
property_valuation    3999 non-null int64
dtypes: int64(3), object(3)
memory usage: 187.5+ KB


In [144]:
cusadd.duplicated().any()


False

In [145]:
cusadd.describe()

customer_id     postcode  property_valuation
count  3999.000000  3999.000000         3999.000000
mean   2003.987997  2985.755939            7.514379
std    1154.576912   844.878364            2.824663
min       1.000000  2000.000000            1.000000
25%    1004.500000  2200.000000            6.000000
50%    2004.000000  2768.000000            8.000000
75%    3003.500000  3750.000000           10.000000
max    4003.000000  4883.000000           12.000000

# <U> 3. INFERENCES </U>
### <U > QUALITY ISSUES </U>
##### `TRANSACTIONS` table
- There are totally 555 rows with atleast one missing value.
- The column 'online_order' contains the most number(360) of missing values followed by other columns(each of these columns having 197 missing values)
  >brand                    
  >product_class              
  >product_size 
  >product_line                
  >standard_cost                
  >product_first_sold_date
- Most of these missing values correspond to the product_id '0'
- The column "product_first_sold_date" datatype has to be changed from float to datetime for analytical purposes  

##### `CUSTOMER DEMOGRAPHIC` table
- Lots of rows with missing values are found. 
- Some column names could be altered into a shorter form for analytical purposes.
- The column 'gender' has Male,Female values and their abbreviations other times.(There are also inconsistent values with spelling errors)
- There is an outlier found in the DOB column that corresponds to an unreasonable age.
- There is an unwanted column named 'default' filled with gibberish and could be removed for better data exploration.

##### `CUSTOMER ADDRESS` table
- Erroneous datatype for the postcode column (It is a string and not an integer)
- Unnecessary zeroes are present in the address column.
- The state column is filled with the abbreviation of the the respective states but sometimes it includes the full-form as well.(Inconsistent value)


In [162]:
all_columns = pd.Series(list(transac) + list(cusadd) + list(cusdem))
all_columns[all_columns.duplicated()]

13    customer_id
19    customer_id
dtype: object

###  `THE KEY JOINING ALL THREE TABLES IS 'CUSTOMER ID'`

# 4. CLEANING 

In [314]:
transac_clean = transac.copy()
cusdem_clean = cusdem.copy()
cusadd_clean = cusadd.copy()

`TRANSACTIONS TABLE`

In [315]:
transac_clean.tail()

transaction_id  product_id  customer_id transaction_date  online_order  \
19995           19996          51         1018       2017-06-24           1.0   
19996           19997          41          127       2017-11-09           1.0   
19997           19998          87         2284       2017-04-14           1.0   
19998           19999           6         2764       2017-07-03           0.0   
19999           20000          11         1144       2017-09-22           1.0   

      order_status          brand product_line product_class product_size  \
19995     Approved     OHM Cycles     Standard          high       medium   
19996     Approved          Solex         Road        medium       medium   
19997     Approved     OHM Cycles     Standard        medium       medium   
19998     Approved     OHM Cycles     Standard          high       medium   
19999     Approved  Trek Bicycles     Standard        medium        small   

       list_price  standard_cost  product_first_sold_date  
19995     2005.66        1203.40                  37823.0  
19996      416.98         312.74                  35560.0  
19997     1636.90          44.71                  40410.0  
19998      227.88         136.73                  38216.0  
19999     1775.81        1580.47                  36334.0

In [316]:
transac_clean.dropna(how ='any',inplace = True)

In [317]:
transac_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19445 entries, 0 to 19999
Data columns (total 13 columns):
transaction_id             19445 non-null int64
product_id                 19445 non-null int64
customer_id                19445 non-null int64
transaction_date           19445 non-null datetime64[ns]
online_order               19445 non-null float64
order_status               19445 non-null object
brand                      19445 non-null object
product_line               19445 non-null object
product_class              19445 non-null object
product_size               19445 non-null object
list_price                 19445 non-null float64
standard_cost              19445 non-null float64
product_first_sold_date    19445 non-null float64
dtypes: datetime64[ns](1), float64(4), int64(3), object(5)
memory usage: 2.1+ MB


In [318]:
final_date = []

In [319]:
import datetime
serial = transac_clean['product_first_sold_date'].tolist()
for i in range(len(serial)):
    seconds = (serial[i] - 25569) * 86400.0
    finaldate = datetime.datetime.utcfromtimestamp(seconds)
    final_date.append(finaldate)

In [320]:
transac_clean['sold_on'] = final_date 

In [321]:
transac_clean.drop(columns =['product_first_sold_date'],inplace = True)

In [384]:
transac_clean.head(3)

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   

  order_status          brand product_line product_class product_size  \
0     Approved          Solex     Standard        medium       medium   
1     Approved  Trek Bicycles     Standard        medium        large   
2     Approved     OHM Cycles     Standard           low       medium   

   list_price  standard_cost    sold_on  
0       71.49          53.62 2012-12-02  
1     2091.47         388.92 2014-03-03  
2     1793.43         248.82 1999-07-20

In [322]:
transac_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19445 entries, 0 to 19999
Data columns (total 13 columns):
transaction_id      19445 non-null int64
product_id          19445 non-null int64
customer_id         19445 non-null int64
transaction_date    19445 non-null datetime64[ns]
online_order        19445 non-null float64
order_status        19445 non-null object
brand               19445 non-null object
product_line        19445 non-null object
product_class       19445 non-null object
product_size        19445 non-null object
list_price          19445 non-null float64
standard_cost       19445 non-null float64
sold_on             19445 non-null datetime64[ns]
dtypes: datetime64[ns](2), float64(3), int64(3), object(5)
memory usage: 2.1+ MB


`CUSTOMER DEMOGRAPHIC`

In [323]:
cusdem_clean.dropna(how='any',inplace= True)

In [324]:
cusdem_clean.head()

customer_id first_name  last_name  gender  \
0            1    Laraine  Medendorp       F   
1            2        Eli    Bockman    Male   
2            3      Arlin     Dearle    Male   
8            9       Mala       Lind  Female   
9           10   Fiorenze    Birdall  Female   

   past_3_years_bike_related_purchases        DOB  \
0                                   93 1953-10-12   
1                                   81 1980-12-16   
2                                   61 1954-01-20   
8                                   97 1973-03-10   
9                                   49 1988-10-11   

                              job_title job_industry_category  \
0                   Executive Secretary                Health   
1                Administrative Officer    Financial Services   
2                    Recruiting Manager              Property   
8  Business Systems Development Analyst           Argiculture   
9               Senior Quality Engineer    Financial Services   

      wealth_segment deceased_indicator                       default  \
0      Mass Customer                  N                            "'   
1      Mass Customer                  N  <script>alert('hi')</script>   
2      Mass Customer                  N           2018-02-01 00:00:00   
8  Affluent Customer                  N                           0/0   
9      Mass Customer                  N                          ð©ð½   

  owns_car  tenure  
0      Yes    11.0  
1      Yes    16.0  
2      Yes    15.0  
8      Yes     8.0  
9      Yes    20.0

In [325]:
cusdem_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2630 entries, 0 to 3996
Data columns (total 13 columns):
customer_id                            2630 non-null int64
first_name                             2630 non-null object
last_name                              2630 non-null object
gender                                 2630 non-null object
past_3_years_bike_related_purchases    2630 non-null int64
DOB                                    2630 non-null datetime64[ns]
job_title                              2630 non-null object
job_industry_category                  2630 non-null object
wealth_segment                         2630 non-null object
deceased_indicator                     2630 non-null object
default                                2630 non-null object
owns_car                               2630 non-null object
tenure                                 2630 non-null float64
dtypes: datetime64[ns](1), float64(1), int64(2), object(9)
memory usage: 287.7+ KB


In [326]:
cusdem['gender'].value_counts()

Female    2037
Male      1872
U           88
M            1
Femal        1
F            1
Name: gender, dtype: int64

In [327]:
cusdem_clean['gender'].value_counts()

gender_abbrev = {'M': 'Male',
                'F': 'Female',
                'Femal': 'Female',
                'U':'Unknown' }
print(gender_abbrev)

cusdem_clean.replace({"gender": gender_abbrev},inplace = True) 

{'M': 'Male', 'F': 'Female', 'Femal': 'Female', 'U': 'Unknown'}


In [328]:
cusdem_clean['gender'].value_counts()

Female    1368
Male      1262
Name: gender, dtype: int64

In [329]:
from datetime import *

now = pd.to_datetime('now')
cusdem_clean['age'] = (now.year - cusdem_clean['DOB'].dt.year) - ((now.month - cusdem_clean['DOB'].dt.month) < 0)

In [330]:
cusdem_clean['age'].sort_values(ascending=False)

719     88
1091    84
657     76
2412    76
0       66
2494    66
2609    66
219     66
229     66
559     66
2598    66
254     66
1814    66
2528    66
3392    66
2485    66
2745    66
2481    66
1914    66
1948    66
2018    66
466     66
983     66
365     66
2177    66
3470    66
370     66
1709    66
424     66
3624    66
        ..
539     19
2807    19
2997    19
3366    19
3014    19
2627    19
2599    19
372     19
2391    19
3615    19
3996    19
828     19
3076    19
1812    19
3111    19
1797    19
3606    19
2758    18
3948    18
1249    18
211     18
1321    18
1518    18
1401    18
244     18
2295    18
65      18
750     18
104     18
446     18
Name: age, Length: 2630, dtype: int64

In [331]:
cusdem_clean.drop(columns =['DOB','default'],inplace = True)

In [332]:
cusdem_clean.tail()

customer_id first_name last_name  gender  \
3992         3993       Andi   Dumelow  Female   
3993         3994    Stephie     Byars  Female   
3994         3995      Rusty    Iapico    Male   
3995         3996    Rosalia  Halgarth  Female   
3996         3997     Blanch    Nisuis  Female   

      past_3_years_bike_related_purchases                     job_title  \
3992                                    6                     Librarian   
3993                                    5  Structural Analysis Engineer   
3994                                   93               Staff Scientist   
3995                                    8         VP Product Management   
3996                                   87               Statistician II   

     job_industry_category     wealth_segment deceased_indicator owns_car  \
3992         Entertainment      Mass Customer                  N       No   
3993         Manufacturing  Affluent Customer                  N       No   
3994         Manufacturing      Mass Customer                  N      Yes   
3995                Health      Mass Customer                  N       No   
3996         Manufacturing     High Net Worth                  N      Yes   

      tenure  age  
3992    10.0   45  
3993    12.0   31  
3994    14.0   44  
3995    19.0   44  
3996     1.0   19

In [333]:
cusdem_clean.rename(columns = {'past_3_years_bike_related_purchases': '3y_bikes_purchased'},inplace = True)

In [334]:
cusdem_clean.rename(columns = {'deceased_indicator': 'D_indicator'},inplace = True)

In [335]:
cusdem_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2630 entries, 0 to 3996
Data columns (total 12 columns):
customer_id              2630 non-null int64
first_name               2630 non-null object
last_name                2630 non-null object
gender                   2630 non-null object
3y_bikes_purchased       2630 non-null int64
job_title                2630 non-null object
job_industry_category    2630 non-null object
wealth_segment           2630 non-null object
D_indicator              2630 non-null object
owns_car                 2630 non-null object
tenure                   2630 non-null float64
age                      2630 non-null int64
dtypes: float64(1), int64(3), object(8)
memory usage: 267.1+ KB


`CUSTOMER ADDRESS`

Erroneous datatype for the postcode column (It is a string and not an integer)
Unnecessary zeroes are present in the address column.
The state column is filled with the abbreviation of the the respective states but sometimes it includes the full-form as well.(Inconsistent value)

In [336]:
cusadd_clean.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

In [337]:
cusadd_clean['postcode'].astype(str);

In [338]:
cusadd_clean['country'].value_counts()


Australia    3999
Name: country, dtype: int64

In [380]:
cusadd['state'].value_counts()

NSW                2054
VIC                 939
QLD                 838
New South Wales      86
Victoria             82
Name: state, dtype: int64

In [339]:
state_abbrev = {'New South Wales': 'NSW',
                'Victoria': 'VIC'}

print(state_abbrev)

cusadd_clean.replace({"state": state_abbrev},inplace = True) 

{'New South Wales': 'NSW', 'Victoria': 'VIC'}


# MERGING 

In [340]:
cusdem_final = cusdem_clean.copy()
cusadd_final = cusadd_clean.copy()
transac_final = transac_clean.copy()

In [341]:
tempdf = pd.merge(cusadd_final, transac_final, on='customer_id')

In [342]:
tempdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19416 entries, 0 to 19415
Data columns (total 18 columns):
customer_id           19416 non-null int64
address               19416 non-null object
postcode              19416 non-null int64
state                 19416 non-null object
country               19416 non-null object
property_valuation    19416 non-null int64
transaction_id        19416 non-null int64
product_id            19416 non-null int64
transaction_date      19416 non-null datetime64[ns]
online_order          19416 non-null float64
order_status          19416 non-null object
brand                 19416 non-null object
product_line          19416 non-null object
product_class         19416 non-null object
product_size          19416 non-null object
list_price            19416 non-null float64
standard_cost         19416 non-null float64
sold_on               19416 non-null datetime64[ns]
dtypes: datetime64[ns](2), float64(3), int64(5), object(8)
memory usage: 2.8+ MB


In [343]:
tempdf.isna().sum()

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
transaction_id        0
product_id            0
transaction_date      0
online_order          0
order_status          0
brand                 0
product_line          0
product_class         0
product_size          0
list_price            0
standard_cost         0
sold_on               0
dtype: int64

In [344]:
tempdf.dropna(how='any',inplace= True)


In [347]:
finaldf = pd.merge(cusdem_final,tempdf,on='customer_id')

In [348]:
finaldf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12970 entries, 0 to 12969
Data columns (total 29 columns):
customer_id              12970 non-null int64
first_name               12970 non-null object
last_name                12970 non-null object
gender                   12970 non-null object
3y_bikes_purchased       12970 non-null int64
job_title                12970 non-null object
job_industry_category    12970 non-null object
wealth_segment           12970 non-null object
D_indicator              12970 non-null object
owns_car                 12970 non-null object
tenure                   12970 non-null float64
age                      12970 non-null int64
address                  12970 non-null object
postcode                 12970 non-null int64
state                    12970 non-null object
country                  12970 non-null object
property_valuation       12970 non-null int64
transaction_id           12970 non-null int64
product_id               12970 non-null int64
tran

In [356]:
finaldf.head()

customer_id first_name  last_name  gender  3y_bikes_purchased  \
0            1    Laraine  Medendorp  Female                  93   
1            1    Laraine  Medendorp  Female                  93   
2            1    Laraine  Medendorp  Female                  93   
3            1    Laraine  Medendorp  Female                  93   
4            1    Laraine  Medendorp  Female                  93   

             job_title job_industry_category wealth_segment D_indicator  \
0  Executive Secretary                Health  Mass Customer           N   
1  Executive Secretary                Health  Mass Customer           N   
2  Executive Secretary                Health  Mass Customer           N   
3  Executive Secretary                Health  Mass Customer           N   
4  Executive Secretary                Health  Mass Customer           N   

  owns_car  ...  transaction_date  online_order order_status           brand  \
0      Yes  ...        2017-12-23           0.0     Approved      OHM Cycles   
1      Yes  ...        2017-04-06           1.0     Approved           Solex   
2      Yes  ...        2017-05-11           1.0     Approved   Trek Bicycles   
3      Yes  ...        2017-01-05           0.0     Approved  Norco Bicycles   
4      Yes  ...        2017-02-21           0.0     Approved           Solex   

  product_line product_class  product_size  list_price  standard_cost  \
0     Standard        medium        medium      235.63         125.07   
1     Standard        medium        medium     1577.53         826.51   
2         Road           low         small     1720.70        1531.42   
3     Standard        medium        medium      360.40         270.30   
4     Standard        medium        medium       71.49          53.62   

     sold_on  
0 2005-05-10  
1 2008-03-19  
2 2003-07-21  
3 2003-09-09  
4 2005-08-09  

[5 rows x 29 columns]

In [360]:
finaldf.set_index('customer_id',inplace =True)

In [379]:
finaldf.sample(7)

first_name last_name  gender  3y_bikes_purchased  \
customer_id                                                     
590               Ddene  Burleton  Female                  85   
100               Tripp     Steed    Male                  80   
2111              Ellis     Vokes    Male                  98   
1226         Constantia      Sirr  Female                  32   
2154               Bird      Bolf  Female                  86   
3406               Lucy  Lackmann  Female                  35   
2936              Pavia    Gethen  Female                  23   

                                        job_title job_industry_category  \
customer_id                                                               
590                           Assistant Professor                Health   
100                               General Manager    Financial Services   
2111                           Speech Pathologist              Property   
1226                            Director of Sales                Health   
2154                            Chemical Engineer         Manufacturing   
3406         Business Systems Development Analyst                    IT   
2936                              Project Manager    Financial Services   

                wealth_segment D_indicator owns_car  tenure  ...  \
customer_id                                                  ...   
590              Mass Customer           N      Yes    11.0  ...   
100              Mass Customer           N      Yes     4.0  ...   
2111         Affluent Customer           N      Yes    12.0  ...   
1226             Mass Customer           N      Yes    13.0  ...   
2154            High Net Worth           N       No     6.0  ...   
3406             Mass Customer           N      Yes    10.0  ...   
2936             Mass Customer           N      Yes    11.0  ...   

             transaction_date online_order  order_status           brand  \
customer_id                                                                
590                2017-11-17          1.0      Approved        WeareA2B   
100                2017-05-10          1.0      Approved      OHM Cycles   
2111               2017-02-11          0.0      Approved           Solex   
1226               2017-06-25          1.0      Approved      OHM Cycles   
2154               2017-08-09          1.0      Approved           Solex   
3406               2017-09-07          1.0      Approved   Trek Bicycles   
2936               2017-07-26          0.0      Approved  Giant Bicycles   

            product_line  product_class  product_size  list_price  \
customer_id                                                         
590             Standard         medium        medium     1292.84   
100             Standard           high        medium     2005.66   
2111            Standard           high        medium      748.17   
1226                Road         medium        medium      742.54   
2154            Standard         medium        medium     1945.43   
3406            Mountain            low        medium      574.64   
2936                Road         medium        medium      792.90   

            standard_cost    sold_on  
customer_id                           
590                 13.44 2003-07-21  
100               1203.40 2014-07-28  
2111               448.90 2006-11-10  
1226               667.40 2016-07-09  
2154               333.18 2002-08-31  
3406               459.71 2003-02-07  
2936               594.68 2003-03-18  

[7 rows x 28 columns]

# The end of data quality and assesment :)